### ISSUE TITLE
updated project-optimizing ml pipeline

### ISSUE THREAD

**Issue by lara-hash** [lara-hash/Optimizing-ML-Pipeline#1]
updated project 1-optimizing ml pipeline

**Comment by lara-hash** [comment]
# Optimizing an ML Pipeline in Azure

## Overview
This project is part of the Udacity Azure ML Nanodegree.
In this project, I built and optimized an Azure ML pipeline by using the Python SDK and a provided Scikit-learn model which resulted in Logistic regression as my best model for both AutoML and HyperDrive. Both models were compared using their metrics and then analyzed.

![image](https://user-images.githubusercontent.com/77818698/107732827-09666600-6d02-11eb-9012-0e827b6dda0f.png)
![image](https://user-images.githubusercontent.com/77818698/107735250-f35ba400-6d07-11eb-9913-f1dacb4e3165.png)
![image](https://user-images.githubusercontent.com/77818698/107735274-fe163900-6d07-11eb-884e-f1c38c8fa7fa.png)


## Summary
**In 1-2 sentences, explain the problem statement: e.g "This dataset contains data about... we seek to predict..."**

I used the bank_marketing dataset which contains 21 columns and 10000 rows. My focus was on the age group and predicting the models through hyperparameter tuning and then comparing them


![image](https://user-images.githubusercontent.com/77818698/107747048-80f6be00-6d1f-11eb-9e4b-caf04bf9d2a0.png)
![image](https://user-images.githubusercontent.com/77818698/107747108-9ff55000-6d1f-11eb-8d67-485b26d7034a.png)


**In 1-2 sentences, explain the solution: e.g. "The best performing model was a ..."**

The best model was chosen after training automl using train.py. I set up the hyperdrive config as well as the compute cluster and then compared both models for accuracy.

![image](https://user-images.githubusercontent.com/77818698/107734646-99a6aa00-6d06-11eb-90e9-c5a5c725f3b3.png)
![image](https://user-images.githubusercontent.com/77818698/107734694-b3e08800-6d06-11eb-881d-d0b8a5377368.png)
![image](https://user-images.githubusercontent.com/77818698/107734757-d4104700-6d06-11eb-8a65-ca6696186586.png)

![image](https://user-images.githubusercontent.com/77818698/107734785-e38f9000-6d06-11eb-854b-78d745a677ed.png)
![image](https://user-images.githubusercontent.com/77818698/107734807-f013e880-6d06-11eb-8c75-dd8689034659.png)
![image](https://user-images.githubusercontent.com/77818698/107734832-fdc96e00-6d06-11eb-800a-4d4ec70c0e4d.png)

## Scikit-learn Pipeline
**Explain the pipeline architecture, including data, hyperparameter tuning, and classification algorithm.**

I decided to choose Random Sampler because it identifies hyperparameter by doing a random selection
![image](https://user-images.githubusercontent.com/77818698/107733595-0b312900-6d04-11eb-994b-8208965afb6e.png)

![image](https://user-images.githubusercontent.com/77818698/107734898-23ef0e00-6d07-11eb-8425-14ff12a0eaa9.png)


![image](https://user-images.githubusercontent.com/77818698/107733509-e0df6b80-6d03-11eb-927f-3ab883aac314.png)

![image](https://user-images.githubusercontent.com/77818698/107733709-53e8e200-6d04-11eb-8116-c4dcd041ccf4.png)


![image](https://user-images.githubusercontent.com/77818698/107733784-85fa4400-6d04-11eb-9cd5-719019c6ee27.png)

**What are the benefits of the parameter sampler you chose?**
![image](https://user-images.githubusercontent.com/77818698/107735787-34a08380-6d09-11eb-8f29-2ed58ae88227.png)
![image](https://user-images.githubusercontent.com/77818698/107735824-4b46da80-6d09-11eb-838b-544b05ca6c96.png)



**What are the benefits of the early stopping policy you chose?**

The runs were set up for 3 hours and I had to adopt early stopping to save time and cost.

## AutoML
**In 1-2 sentences, describe the model and hyperparameters generated by AutoML.**

Attached below the VotingEnsemble classifier used during my runs. With an accuracy of 90% from 25 models from AutoML and Hyperdrive.

![image](https://user-images.githubusercontent.com/77818698/107749283-0039c100-6d23-11eb-88b2-bfd91bd1659b.png)
![image](https://user-images.githubusercontent.com/77818698/107749311-0af45600-6d23-11eb-9b09-b0ab54001230.png)


![image](https://user-images.githubusercontent.com/77818698/107748757-3a569300-6d22-11eb-9af1-cc645c6e224c.png)


![image](https://user-images.githubusercontent.com/77818698/107734941-3e28ec00-6d07-11eb-96d8-64a11a920f72.png)

![image](https://user-images.githubusercontent.com/77818698/107734982-50a32580-6d07-11eb-86ca-689f7abe6cef.png)

## Pipeline comparison
**Compare the two models and their performance. What are the differences in accuracy? In architecture? If there was a difference, why do you think there was one?**
![image](https://user-images.githubusercontent.com/77818698/107735384-3cabf380-6d08-11eb-8021-1bcfd50006ff.png)
![image](https://user-images.githubusercontent.com/77818698/107735464-6bc26500-6d08-11eb-9103-f748ed6cd6cf.png)

![image](https://user-images.githubusercontent.com/77818698/107735042-72041180-6d07-11eb-97b8-1f63d2b37f67.png)
![image](https://user-images.githubusercontent.com/77818698/107735080-8b0cc280-6d07-11eb-948b-6d4bebcc6cfe.png)
![image](https://user-images.githubusercontent.com/77818698/107735103-9bbd3880-6d07-11eb-9b2f-377bae1287a7.png)
![image](https://user-images.githubusercontent.com/77818698/107735126-aa0b5480-6d07-11eb-9153-54be6c9536ac.png)
![image](https://user-images.githubusercontent.com/77818698/107735142-b5f71680-6d07-11eb-8824-d5cff3c5a8fa.png)

## Future work
An improvement will be achieved by choosing a different set of parameters and metrics. This will result in a different outcome in metrics like weighted AUC, Precision, Recall etc. Below is the result from my runs.

![image](https://user-images.githubusercontent.com/77818698/107734126-5b5cbb00-6d05-11eb-91bc-4f2737b68175.png)

![image](https://user-images.githubusercontent.com/77818698/107734413-09686500-6d06-11eb-99d7-997683a34a01.png)
![image](https://user-images.githubusercontent.com/77818698/107869438-ace87f80-6e96-11eb-9871-24f73be11637.png)


![image](https://user-images.githubusercontent.com/77818698/107734310-c1e1d900-6d05-11eb-966c-34b92e7f6bdb.png)

![image](https://user-images.githubusercontent.com/77818698/107734378-edfd5a00-6d05-11eb-946e-707df26887af.png)


## Proof of cluster clean up
I did not delete my cluster and left it as it is on the workspace. Attached image below

![image](https://user-images.githubusercontent.com/77818698/107733851-ade9a780-6d04-11eb-8ca8-495663531358.png)

![image](https://user-images.githubusercontent.com/77818698/107734553-5e0be000-6d06-11eb-8ab6-96362bfa9358.png)
![image](https://user-images.githubusercontent.com/77818698/107735478-7846bd80-6d08-11eb-9d76-c9453a6f805d.png)
![image](https://user-images.githubusercontent.com/77818698/107735499-83015280-6d08-11eb-8cee-2f64d0bf7fc1.png)
